### Importing Packages

In [3]:
import pandas as pd
import numpy as np
from math import pow
from functools import partial
from sklearn.metrics import mean_absolute_error, mean_squared_error
import matplotlib.pyplot as plt
from matplotlib.pyplot import figure
import haversine

### Loading Data

In [4]:
events_data = pd.read_csv("./events_3.csv")
location_data = pd.read_csv("./iata.csv")

In [5]:
events_data.head()

ts event_type user_id   date_from     date_to origin  \
0  2017-04-27 11:06:51     search  60225f  2017-06-01  2017-06-07    PAR   
1  2017-04-27 20:15:27       book  e5d69e  2017-08-12  2017-09-02    FRA   
2  2017-04-27 23:03:43       book  f953f0  2017-10-08  2017-10-11    BER   
3  2017-04-27 15:17:50       book  794d35  2017-04-28  2017-05-01    BER   
4  2017-04-27 22:51:57       book  ca4f94  2017-05-16  2017-05-22    DEL   

  destination  num_adults  num_children  
0         NYC           6             1  
1         WAS           3             1  
2         CGN           2             0  
3         BCN           1             0  
4         BKK           4             0

In [6]:
location_data.head()

iata_code    lat    lon
0       LON  51.50  -0.17
1       MOW  55.75  37.70
2       NYC  40.71 -74.01
3       PAR  48.85   2.35
4       IST  41.01  28.95

### Data Sanity Check

In [7]:
location_data.groupby("iata_code").count()

lat  lon
iata_code          
ABV          2    2
ADB          1    1
ADD          2    2
ADL          2    2
AEP          1    1
AER          2    2
AGP          2    2
AKL          2    2
ALA          2    2
ALC          2    2
ALG          2    2
AMS          2    2
ANC          2    2
ANK          1    1
ARN          1    1
ATH          2    2
ATL          2    2
AUH          2    2
AUS          2    2
AYT          2    2
BAH          2    2
BCN          2    2
BER          1    1
BET          2    2
BGY          1    1
BHX          2    2
BHZ          1    1
BJS          1    1
BKK          2    2
BLR          2    2
...        ...  ...
UIO          2    2
UPG          2    2
URC          3    3
VCE          2    2
VCP          1    1
VIE          2    2
VKO          1    1
VLI          2    2
WAS          1    1
WAW          2    2
WIL          1    1
WLG          2    2
WUH          2    2
XIY          1    1
XMN          2    2
YEA          1    1
YEG          1    1
YKS          2    2
YMQ          1    1
YOW          2    2
YTO          1    1
YUL          1    1
YVR          2    2
YWG          2    2
YXL          2    2
YYC          2    2
YYZ          1    1
YZF          2    2
ZME          1    1
ZRH          2    2

[298 rows x 2 columns]

In IATA Codes multiple locations are given for many codes. Dropping the duplicates and keeping only the first location for the sake of data sanity.

In [8]:
location_data.drop_duplicates(subset= 'iata_code', inplace=True)

In [9]:
events_data.drop_duplicates(inplace=True)

### Merging DataFrames

In [10]:
events_data[events_data.user_id == 'f953f0']

ts event_type user_id   date_from     date_to origin  \
2     2017-04-27 23:03:43       book  f953f0  2017-10-08  2017-10-11    BER   
134   2017-04-27 22:40:42     search  f953f0  2017-10-08  2017-10-11    BER   
359   2017-04-27 21:57:56     search  f953f0  2017-10-08  2017-10-11    BER   
9688  2017-04-27 21:56:36     search  f953f0  2017-10-08  2017-10-11    BER   

     destination  num_adults  num_children  
2            CGN           2             0  
134          CGN           2             0  
359          CGN           2             0  
9688         CGN           2             0

In [11]:
events_origin_loc = events_data.merge(location_data, left_on='origin', right_on='iata_code', how = "left")
events_origin_loc.rename(columns={'lat' : 'origin_lat', "lon" : "origin_lon"}, inplace=True)
events_origin_loc.head()

ts event_type user_id   date_from     date_to origin  \
0  2017-04-27 11:06:51     search  60225f  2017-06-01  2017-06-07    PAR   
1  2017-04-27 20:15:27       book  e5d69e  2017-08-12  2017-09-02    FRA   
2  2017-04-27 23:03:43       book  f953f0  2017-10-08  2017-10-11    BER   
3  2017-04-27 15:17:50       book  794d35  2017-04-28  2017-05-01    BER   
4  2017-04-27 22:51:57       book  ca4f94  2017-05-16  2017-05-22    DEL   

  destination  num_adults  num_children iata_code  origin_lat  origin_lon  
0         NYC           6             1       PAR       48.85        2.35  
1         WAS           3             1       FRA       50.12        8.68  
2         CGN           2             0       BER       52.52       13.41  
3         BCN           1             0       BER       52.52       13.41  
4         BKK           4             0       DEL       28.57       77.10

In [12]:
events_origin_loc[events_origin_loc.user_id == 'f953f0']

ts event_type user_id   date_from     date_to origin  \
2     2017-04-27 23:03:43       book  f953f0  2017-10-08  2017-10-11    BER   
134   2017-04-27 22:40:42     search  f953f0  2017-10-08  2017-10-11    BER   
358   2017-04-27 21:57:56     search  f953f0  2017-10-08  2017-10-11    BER   
9260  2017-04-27 21:56:36     search  f953f0  2017-10-08  2017-10-11    BER   

     destination  num_adults  num_children iata_code  origin_lat  origin_lon  
2            CGN           2             0       BER       52.52       13.41  
134          CGN           2             0       BER       52.52       13.41  
358          CGN           2             0       BER       52.52       13.41  
9260         CGN           2             0       BER       52.52       13.41

In [13]:
events_location = events_origin_loc.merge(location_data, left_on='destination', right_on='iata_code', how = "left")
events_location.rename(columns={'lat' : 'dest_lat', "lon" : "dest_lon"}, inplace=True)
events_location.head()

ts event_type user_id   date_from     date_to origin  \
0  2017-04-27 11:06:51     search  60225f  2017-06-01  2017-06-07    PAR   
1  2017-04-27 20:15:27       book  e5d69e  2017-08-12  2017-09-02    FRA   
2  2017-04-27 23:03:43       book  f953f0  2017-10-08  2017-10-11    BER   
3  2017-04-27 15:17:50       book  794d35  2017-04-28  2017-05-01    BER   
4  2017-04-27 22:51:57       book  ca4f94  2017-05-16  2017-05-22    DEL   

  destination  num_adults  num_children iata_code_x  origin_lat  origin_lon  \
0         NYC           6             1         PAR       48.85        2.35   
1         WAS           3             1         FRA       50.12        8.68   
2         CGN           2             0         BER       52.52       13.41   
3         BCN           1             0         BER       52.52       13.41   
4         BKK           4             0         DEL       28.57       77.10   

  iata_code_y  dest_lat  dest_lon  
0         NYC     40.71    -74.01  
1         WAS     38.90    -77.04  
2         CGN     50.88      7.12  
3         BCN     41.39      2.16  
4         BKK     13.75    100.50

In [12]:
events_location[events_location.user_id == 'f953f0']

ts event_type user_id   date_from     date_to origin  \
2     2017-04-27 23:03:43       book  f953f0  2017-10-08  2017-10-11    BER   
134   2017-04-27 22:40:42     search  f953f0  2017-10-08  2017-10-11    BER   
358   2017-04-27 21:57:56     search  f953f0  2017-10-08  2017-10-11    BER   
9260  2017-04-27 21:56:36     search  f953f0  2017-10-08  2017-10-11    BER   

     destination  num_adults  num_children iata_code_x  origin_lat  \
2            CGN           2             0         BER       52.52   
134          CGN           2             0         BER       52.52   
358          CGN           2             0         BER       52.52   
9260         CGN           2             0         BER       52.52   

      origin_lon iata_code_y  dest_lat  dest_lon  
2          13.41         CGN     50.88      7.12  
134        13.41         CGN     50.88      7.12  
358        13.41         CGN     50.88      7.12  
9260       13.41         CGN     50.88      7.12

### Calculating Distances

In [14]:
calc_distance = lambda x: haversine.haversine((x[0], x[1]), (x[2], x[3]))

events_location.loc[:,'distance_km'] = events_location[['origin_lat', 'origin_lon', 'dest_lat', 'dest_lon']].apply(calc_distance, axis=1) 

In [15]:
events_location.head()

ts event_type user_id   date_from     date_to origin  \
0  2017-04-27 11:06:51     search  60225f  2017-06-01  2017-06-07    PAR   
1  2017-04-27 20:15:27       book  e5d69e  2017-08-12  2017-09-02    FRA   
2  2017-04-27 23:03:43       book  f953f0  2017-10-08  2017-10-11    BER   
3  2017-04-27 15:17:50       book  794d35  2017-04-28  2017-05-01    BER   
4  2017-04-27 22:51:57       book  ca4f94  2017-05-16  2017-05-22    DEL   

  destination  num_adults  num_children iata_code_x  origin_lat  origin_lon  \
0         NYC           6             1         PAR       48.85        2.35   
1         WAS           3             1         FRA       50.12        8.68   
2         CGN           2             0         BER       52.52       13.41   
3         BCN           1             0         BER       52.52       13.41   
4         BKK           4             0         DEL       28.57       77.10   

  iata_code_y  dest_lat  dest_lon  distance_km  
0         NYC     40.71    -74.01  5837.828026  
1         WAS     38.90    -77.04  6530.035012  
2         CGN     50.88      7.12   470.077408  
3         BCN     41.39      2.16  1499.761224  
4         BKK     13.75    100.50  2923.178366

### Data Understanding

In [23]:
events_data.event_type.unique()

array(['search', 'book'], dtype=object)

There are two types of events i.e., Search and Book

In [40]:
events_location.shape,\
events_location[events_location.event_type == 'book'].shape,\
events_location[events_location.event_type == 'search'].shape

((45898, 18), (1808, 18), (44090, 18))

Total Events = 45898
Total Searches = 44,090
Total Bookings = 1,808

In [16]:
events_location.loc[:,'ts_dt'] = pd.to_datetime(events_location.ts)
events_location.ts_dt.max() - events_location.ts_dt.min()

Timedelta('13 days 23:25:25')

The events data is spanned over around 14 days. 

In [63]:
len(events_location.user_id.unique())

31179

There are 31,179 unique users

In [27]:
user_booking_counts = events_location[events_location.event_type=='book'].groupby("user_id").agg({'event_type': 'count'})
user_booking_counts.reset_index(inplace=True)
user_booking_counts[user_booking_counts.event_type > 1]

user_id  event_type
52   08e914           2
734  690a8c           3

Only 2 users have done booking more than once.

In [28]:
uids = user_booking_counts[user_booking_counts.event_type > 1].user_id
events_location[events_location.event_type == 'book'][events_location.user_id.isin(uids.values)]

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


ts event_type user_id   date_from     date_to origin  \
4625  2017-04-21 02:43:58       book  690a8c  2017-07-12  2017-07-12    BNE   
4767  2017-04-21 02:43:51       book  690a8c  2017-07-12  2017-07-12    BNE   
6555  2017-04-26 00:57:37       book  08e914  2017-04-27  2017-04-28    DUS   
8438  2017-04-28 18:34:01       book  08e914  2017-05-03  2017-05-03    DUS   
8712  2017-04-27 23:52:13       book  690a8c  2017-05-10  2017-05-10    MUC   

     destination  num_adults  num_children iata_code_x  origin_lat  \
4625         SIN           2             2         BNE      -27.47   
4767         SIN           2             2         BNE      -27.47   
6555         VIE           1             0         DUS       51.22   
8438         VIE           1             0         DUS       51.22   
8712         DUS           1             0         MUC       48.14   

      origin_lon iata_code_y  dest_lat  dest_lon  distance_km  \
4625      153.03         SIN      1.29    103.85  6147.753956   
4767      153.03         SIN      1.29    103.85  6147.753956   
6555        6.78         VIE     48.12     16.56   782.845564   
8438        6.78         VIE     48.12     16.56   782.845564   
8712       11.58         DUS     51.22      6.78   486.185022   

                   ts_dt  
4625 2017-04-21 02:43:58  
4767 2017-04-21 02:43:51  
6555 2017-04-26 00:57:37  
8438 2017-04-28 18:34:01  
8712 2017-04-27 23:52:13

In [60]:
events_location.loc[:,'trip'] = events_location[['origin', 'destination']].apply(lambda x : x[0] + "_" + x[1], axis= 1)
events_location.loc[:,'dt_from'] = pd.to_datetime(events_location.date_from)
events_location.loc[:,'dt_to'] = pd.to_datetime(events_location.date_to)
events_location.loc[:,'duration'] = events_location[['dt_to', 'dt_from']].apply(lambda x : x[0] - x[1], axis =1)
events_location.loc[:,'from_to'] = events_location[['date_from', 'date_to']].apply(lambda x : str(x[0]) + "_" + str(x[1]), axis= 1)

In [70]:
len(events_location[['from_to', 'trip', 'num_adults', 'num_children']].drop_duplicates()),\
len(events_location[events_location.event_type == 'book'][['from_to', 'trip', 'num_adults', 'num_children']].drop_duplicates())

(35787, 1801)

In [71]:
1801 / 35787

0.05032553720624808

There are only 35,787 unique search/book criteria, out of which 1801 are booked. There is only 5% success rate for every unique search.

### Creating 'search and book' timeline for each user and trip

In [72]:
event_timeline = events_location.groupby(["user_id", 'trip', 'from_to', 'distance_km', 'duration']).\
agg({'ts_dt':list, 'event_type': list, 'num_adults' : list, 'num_children' : list})


In [73]:
event_timeline.reset_index(inplace=True)
event_timeline.head()

user_id     trip                from_to  distance_km duration  \
0  000069  FRA_ALG  2017-05-10_2017-05-26  1553.717725  16 days   
1  00019d  PAR_SSA  2017-09-15_2017-10-13  7944.570036  28 days   
2  0002e3  PAR_ATH  2017-05-01_2017-05-01  2095.453304   0 days   
3  000332  NCE_TLV  2017-06-08_2017-07-07  2722.329760  29 days   
4  000332  TLV_NCE  2017-06-08_2017-07-07  2722.329760  29 days   

                                               ts_dt  \
0                              [2017-04-29 17:59:51]   
1                              [2017-04-28 19:14:23]   
2                              [2017-04-24 21:25:10]   
3  [2017-04-27 16:35:35, 2017-04-27 16:34:02, 201...   
4                              [2017-04-27 21:24:01]   

                 event_type num_adults num_children  
0                  [search]        [2]          [3]  
1                  [search]        [4]          [0]  
2                  [search]        [1]          [0]  
3  [search, search, search]  [1, 1, 2]    [0, 0, 0]  
4                  [search]        [2]          [0]

In [150]:
booking_timelines = event_timeline[event_timeline.event_type.map(lambda x : 'book' in x)]
non_booking_timelines = event_timeline[event_timeline.event_type.map(lambda x : 'book' not in x)]

In [151]:
booking_timelines.loc[:, 'ts_dt_sorted'] = booking_timelines.ts_dt.apply(sorted)
non_booking_timelines.loc[:, 'ts_dt_sorted'] = non_booking_timelines.ts_dt.apply(sorted)

In [133]:
booking_timelines

user_id     trip                from_to   distance_km duration  \
31     0035d4  DUS_AGP  2018-02-08_2018-02-20   1839.949396  12 days   
42     004671  DUS_MUC  2017-05-14_2017-05-15    486.185022   1 days   
49     005e14  PAR_YMQ  2017-05-07_2017-05-20   5506.315396  13 days   
57     006e35  HAM_TXL  2017-08-23_2017-08-31    246.123282   8 days   
165    011d58  BER_FRA  2017-04-28_2017-04-28    423.247353   0 days   
208    0166cf  MUC_PMI  2017-07-06_2017-07-15   1188.284804   9 days   
218    0182d0  DUS_KTM  2017-06-16_2017-07-05   6871.855604  19 days   
287    01f2a2  CGN_BER  2017-05-19_2017-05-21    470.077408   2 days   
329    022b2f  STR_ANK  2017-08-03_2017-08-17   2112.058125  14 days   
374    027919  BER_AGP  2017-10-12_2017-11-02   2240.628843  21 days   
385    028cd3  BKK_HAN  2017-05-09_2017-05-09   1003.240844   0 days   
388    028fcb  MUC_LIS  2017-09-24_2017-10-09   1963.732768  15 days   
398    02a312  NYC_CUN  2017-05-04_2017-05-12   2501.989890   8 days   
405    02b7c2  PAR_TLS  2017-05-18_2017-05-20    587.944045   2 days   
411    02c13e  DUS_ALC  2017-06-05_2017-06-17   1549.329193  12 days   
421    02cd28  BER_VIE  2017-04-22_2017-04-22    537.820574   0 days   
426    02d63d  NCE_BCN  2017-06-07_2017-06-09    485.309913   2 days   
458    031d5e  FRA_TXL  2017-05-14_2017-05-17    419.565351   3 days   
504    0388d0  HAV_MUC  2017-05-16_2017-05-16   8409.945169   0 days   
534    03c4d2  MRS_TYO  2017-11-27_2017-12-12  10084.352389  15 days   
585    040016  STR_HAM  2017-05-15_2017-05-22    536.927927   7 days   
588    0403b3  PAR_VCE  2017-08-10_2017-08-17    843.760149   7 days   
611    0426b7  MUC_VCE  2017-07-17_2017-07-17    305.603883   0 days   
650    046c18  MUC_LIS  2017-05-06_2017-05-13   1963.732768   7 days   
655    047070  BER_LED  2017-04-22_2017-05-01   1314.268755   9 days   
689    04abef  ZRH_BCN  2017-05-20_2017-05-24    843.412411   4 days   
709    04daae  DUS_AYT  2017-06-01_2017-06-10   2473.829620   9 days   
731    04ffa4  DUS_ZRH  2017-05-17_2017-05-19    438.562348   2 days   
780    057477  STR_MAD  2017-06-27_2017-07-01   1376.462502   4 days   
805    058fb2  DUS_MUC  2017-05-16_2017-05-22    486.185022   6 days   
...       ...      ...                    ...           ...      ...   
35555  fae688  DUS_BKK  2017-04-29_2017-04-29   9075.229329   0 days   
35558  fae7a7  HAM_MUC  2017-08-10_2017-08-11    614.690248   1 days   
35625  fb527d  MUC_DUS  2017-05-31_2017-05-31    486.185022   0 days   
35718  fbfb41  NCE_YVR  2017-09-24_2017-10-03   8611.903185   9 days   
35730  fc16cc  PAR_BCN  2017-05-25_2017-05-26    829.648484   1 days   
35767  fc58ae  DUS_BCN  2017-06-12_2017-06-26   1148.623804  14 days   
35771  fc5df8  DUS_LIS  2017-10-26_2017-11-09   1862.399590  14 days   
35780  fc69c9  FRA_NYC  2017-11-14_2017-11-30   6202.585984  16 days   
35804  fc941a  HAM_LON  2017-07-23_2017-07-28    726.159903   5 days   
35814  fc99b0  MUC_ARN  2017-06-21_2017-06-26   1344.107401   5 days   
35820  fca078  STR_TXL  2017-07-12_2017-07-15    510.269063   3 days   
35827  fcb6a2  BER_JNB  2017-05-19_2017-06-02   8868.076951  14 days   
35842  fcceb0  BER_OSL  2017-05-15_2017-05-21    837.856053   6 days   
35847  fcceb1  FRA_CTU  2017-05-31_2017-06-14   7799.372916  14 days   
35869  fcf1b8  STR_LON  2017-05-19_2017-05-24    731.070713   5 days   
35913  fd49f2  DUS_ZRH  2017-08-08_2017-08-11    438.562348   3 days   
35928  fd6d18  FRA_SGN  2018-01-18_2018-02-03   9645.902446  16 days   
35930  fd6fb5  PAR_PMI  2017-05-08_2017-05-09   1034.554088   1 days   
35933  fd714d  CPT_FRA  2017-11-29_2017-11-29   9400.189402   0 days   
35940  fd7cea  NCE_ORY  2017-04-27_2017-04-28    676.256264   1 days   
35970  fdb61d  FRA_ICN  2017-10-15_2017-11-23   8531.625549  39 days   
36031  fe1fbf  STR_BER  2017-06-03_2017-06-05    511.559243   2 days   
36085  feab36  MUC_BER  2017-05-16_2017-05-24    504.015245   8 days   
36101  febf52  CGN_PRG  2017-10-14_2017-10-17 

In [152]:
non_booking_timelines

user_id     trip                from_to   distance_km duration  \
0      000069  FRA_ALG  2017-05-10_2017-05-26   1553.717725  16 days   
1      00019d  PAR_SSA  2017-09-15_2017-10-13   7944.570036  28 days   
2      0002e3  PAR_ATH  2017-05-01_2017-05-01   2095.453304   0 days   
3      000332  NCE_TLV  2017-06-08_2017-07-07   2722.329760  29 days   
4      000332  TLV_NCE  2017-06-08_2017-07-07   2722.329760  29 days   
5      0004e1  MUC_ROM  2017-07-19_2017-07-25    698.788332   6 days   
6      0004f1  LYS_YUL  2017-09-15_2017-10-05   5849.529555  20 days   
7      0006d6  CDG_TLV  2017-04-30_2017-05-22   3271.590472  22 days   
8      000a42  BKK_HKT  2017-07-21_2017-07-26    670.747014   5 days   
9      000a72  LYS_ALG  2017-05-07_2017-06-17   1012.053485  41 days   
10     000bf6  BER_PAR  2017-04-29_2017-04-29    878.299203   0 days   
11     000c4e  DUS_OPO  2017-09-17_2017-10-04   1624.100628  17 days   
12     000fa9  TLS_VCE  2017-06-26_2017-06-30    886.536342   4 days   
13     00101a  YUL_TXL  2017-09-06_2017-09-14   6003.280477   8 days   
14     001328  BER_MAD  2017-05-13_2017-05-15   1868.931836   2 days   
15     001861  MUC_BKK  2018-02-08_2018-02-22   8794.589830  14 days   
16     001a18  TLS_CAS  2017-05-05_2017-05-05   1361.035496   0 days   
17     001c62  PAR_OPO  2017-09-15_2017-09-15   1212.471145   0 days   
18     001d12  STR_AYT  2017-12-15_2017-12-27   2189.823553  12 days   
19     001e07  BER_ALA  2017-08-06_2017-09-10   4687.942823  35 days   
20     001e67  DUS_VIE  2017-06-21_2017-06-23    782.845564   2 days   
21     001ec2  DUS_MIA  2017-07-20_2017-07-27   7601.564827   7 days   
22     0026bc  AMS_HEL  2017-09-23_2017-09-28   1512.925379   5 days   
23     0026e2  PAR_NYC  2017-07-28_2017-08-11   5837.828026  14 days   
24     00282c  LYS_DEL  2017-09-28_2017-10-19   6453.316963  21 days   
25     00295c  MUC_ATH  2017-06-07_2017-06-09   1496.045840   2 days   
26     002d02  BER_PAR  2017-07-12_2017-07-15    878.299203   3 days   
27     002fc6  TLS_CDG  2017-07-06_2017-07-08    607.558903   2 days   
28     003463  BER_VIE  2017-05-01_2017-05-01    537.820574   0 days   
29     00349c  DUS_BUD  2017-06-10_2017-06-12    995.589318   2 days   
...       ...      ...                    ...           ...      ...   
36249  ffc202  BWI_STR  2017-07-30_2017-08-13   6581.845244  14 days   
36250  ffc6ed  IEV_DUS  2017-12-14_2017-12-18   1658.423563   4 days   
36251  ffc6f8  ZRH_ATH  2017-07-05_2017-07-05   1620.587329   0 days   
36252  ffc6f8  ZRH_ATH  2017-07-07_2017-07-07   1620.587329   0 days   
36253  ffc974  FRA_IZM  2017-04-29_2017-04-29   1953.983046   0 days   
36254  ffc9a9  TLS_LED  2017-10-25_2017-11-01   2643.052569   7 days   
36255  ffcc37  CGN_MXP  2017-08-02_2017-08-05    595.631571   3 days   
36256  ffcc37  CGN_MXP  2017-08-09_2017-08-12    595.631571   3 days   
36257  ffcf92  NCE_LIS  2017-09-29_2017-10-06   1471.108022   7 days   
36258  ffd21d  BER_ZRH  2017-04-30_2017-05-03    661.523074   3 days   
36259  ffd21d  BER_ZRH  2017-05-01_2017-05-04    661.523074   3 days   
36260  ffd46e  PAR_MAD  2017-06-21_2017-06-25   1051.716307   4 days   
36261  ffd48b  BCN_DUS  2017-04-28_2017-04-28   1148.623804   0 days   
36263  ffdb0b  MUC_ATH  2017-05-01_2017-05-06   1496.045840   5 days   
36264  ffdbec  FRA_LIS  2017-05-18_2017-05-23   1890.489032   5 days   
36265  ffdcc4  FRA_CKG  2017-06-03_2017-06-07   8051.575020   4 days   
36266  ffdd2f  FRA_OSL  2017-05-06_2017-05-12   1096.431753   6 days   
36267  ffe0a4  TXL_BHX  2017-06-21_2017-06-24   1025.901367   3 days   
36268  ffe1ec  PAR_ANK  2017-05-08_2017-06-06   2597.273973  29 days   
36269  ffe752  STR_YTO  2017-07-04_2017-07-17   6438.854922  13 days   
36270  ffea87  CGN_TLV  2017-08-23_2017-08-29   3077.913059   6 days   
36271  ffed46  TXL_MUC  2017-05-17_2017-05-17    506.197487   0 days   
36272  fff5ec  HAM_PMI  2017-05-27_2017-05-30   1654.515799   3 days   
36273  fff6a8  BER_OSL  2017-06-11_2017-06-26 

In [155]:
sort_by_key = lambda x : [y for _, y in sorted(zip(x[0], x[1]))]
booking_timelines.loc[:, 'event_type_sorted'] = booking_timelines[['ts_dt', 'event_type']].apply(sort_by_key, axis = 1, raw = True)
booking_timelines.loc[:, 'num_adults_sorted'] = booking_timelines[['ts_dt', 'num_adults']].apply(sort_by_key, axis = 1, raw = True)
booking_timelines.loc[:, 'num_children_sorted'] = booking_timelines[['ts_dt', 'num_children']].apply(sort_by_key, axis = 1, raw = True)

/anaconda3/lib/python3.6/site-packages/pandas/core/indexing.py:362: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[key] = _infer_fill_value(value)
/anaconda3/lib/python3.6/site-packages/pandas/core/indexing.py:543: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [157]:
last_is_book = lambda x : 1 if x[-1] == 'book' else 0 
booking_timelines.loc[:,'last_event_book'] = booking_timelines.event_type_sorted.apply(last_is_book)

In [160]:
booking_timelines[booking_timelines.last_event_book == 0]

user_id     trip                from_to   distance_km duration  \
907    062ea8  HAM_VIE  2017-05-04_2017-05-05    760.513960   1 days   
2051   0ea441  MUC_ATH  2017-09-16_2017-10-18   1496.045840  32 days   
2257   102f6f  MUC_IZM  2017-10-17_2017-10-17   1654.949302   0 days   
2293   106a19  DUS_MOW  2017-05-25_2017-06-04   2087.872994  10 days   
3008   158b7a  BKK_SGN  2017-09-08_2017-09-08    743.299798   0 days   
3409   185fb1  BER_LIS  2017-06-23_2017-07-06   2312.073474  13 days   
3599   19bfe4  MUC_SYD  2017-06-01_2017-06-22  16331.855061  21 days   
3668   1a281d  PAR_CMN  2017-08-17_2017-09-07   1908.325843  21 days   
3736   1ab0c1  MUC_DPS  2017-09-02_2017-09-29  11737.924366  27 days   
4021   1c9e6a  DUS_PMI  2017-11-19_2017-11-26   1335.099389   7 days   
5081   23ebdd  JKT_DPS  2017-08-20_2017-09-01    952.403535  12 days   
5100   240855  BER_NYC  2017-06-27_2017-07-04   6385.775230   7 days   
5396   25ded7  MRS_PRG  2017-06-02_2017-06-05   1010.258640   3 days   
5756   285515  DUS_IST  2017-04-23_2017-05-03   2038.982852  10 days   
6153   2aed06  BER_PMI  2017-08-28_2017-09-11   1657.197524  14 days   
6718   2f0b9b  PAR_ORY  2017-05-04_2017-05-20     13.363504  16 days   
6771   2f543a  CDG_AGP  2017-10-23_2017-11-08   1478.386617  16 days   
6910   3020c8  HAM_LIS  2017-07-14_2017-07-21   2200.875572   7 days   
6974   308ea5  SXF_BCN  2017-04-30_2017-05-06   1492.410172   6 days   
7321   330bee  STR_AMS  2017-05-16_2017-05-19    501.418271   3 days   
7705   35da03  DUS_PMI  2017-09-28_2017-10-05   1335.099389   7 days   
7736   360431  PAR_YMQ  2017-09-26_2017-10-06   5506.315396  10 days   
8258   398421  DUS_AGP  2017-11-14_2017-11-26   1839.949396  12 days   
8283   39b927  FRA_DXB  2017-06-25_2017-07-01   4840.303809   6 days   
8308   39dc7a  TXL_PER  2018-03-30_2018-04-21  13584.644553  22 days   
8319   39e270  BER_PMI  2017-08-28_2017-09-04   1657.197524   7 days   
8334   39fcda  CDG_NCE  2017-06-07_2017-06-10    694.266360   3 days   
9217   402878  MUC_ROM  2017-07-02_2017-07-05    698.788332   3 days   
9311   40d677  DUS_TLV  2017-05-10_2017-05-20   3118.161695  10 days   
9404   419784  STR_HAM  2017-06-14_2017-06-17    536.927927   3 days   
...       ...      ...                    ...           ...      ...   
27921  c4f4ee  PAR_YMQ  2017-06-16_2017-06-26   5506.315396  10 days   
28674  ca4f94  DEL_BKK  2017-05-16_2017-05-22   2923.178366   6 days   
28694  ca7672  DUS_PMI  2017-09-19_2017-09-30   1335.099389  11 days   
28902  cbeb65  DUS_DUB  2017-06-14_2017-06-17    915.597386   3 days   
29157  cdbf38  IST_STR  2017-08-05_2017-08-26   1773.365352  21 days   
29493  d03203  TXL_BKK  2017-07-19_2017-07-31   8610.935922  12 days   
29753  d2152c  DUS_AYT  2017-07-19_2017-08-09   2473.829620  21 days   
29767  d22d91  DUS_SYD  2017-11-08_2017-11-17  16576.042535   9 days   
29802  d2708c  FRA_ICN  2017-09-01_2017-09-27   8531.625549  26 days   
29881  d30e3e  FRA_HEL  2017-07-17_2017-07-20   1513.946038   3 days   
30217  d56a28  FRA_STO  2017-05-05_2017-05-05   1186.219850   0 days   
30374  d6a1e5  DUS_PAR  2017-04-30_2017-04-30    411.649770   0 days   
30487  d77436  ORY_ATH  2017-09-30_2017-10-07   2088.727229   7 days   
30975  dae00a  CGN_CPH  2017-05-17_2017-05-22    642.717862   5 days   
31194  dc380e  PAR_KTM  2017-06-08_2017-06-24   7236.753016  16 days   
31489  de614b  FRA_BKK  2017-04-23_2017-05-07   8966.898802  14 days   
32151  e31200  FRA_BKK  2017-09-13_2017-10-12   8966.898802  29 days   
32293  e4269a  DMK_HKT  2017-05-19_2017-05-19    691.236486   0 days   
32656  e6baf6  BER_MEL  2017-05-02_2017-10-07  15972.494645 158 days   
32745  e77fbc  FRA_NYC  2017-12-17_2017-12-28   6202.585984  11 days   
32782  e7c086  CGN_BER  2017-07-18_2017-07-22    470.077408   4 days   
33192  eab53b  ZRH_LAS  2017-05-02_2017-05-09   9164.832652   7 days   
33206  eabf04  PAR_UIO  2017-06-20_2017-11-20   9360.489828 153 days   
33684  ee3529  DUS_BER  2017-05-21_2017-05-24 

In [162]:
booking_timelines.shape, booking_timelines[booking_timelines.last_event_book == 0].shape

((1806, 14), (137, 14))

Out of all 1806 bookings 137 times users search again after booking and rest do not have any more events after search.

In [165]:
from collections import Counter
book_count = lambda x : Counter(x)['book']
booking_timelines.loc[:,'num_book'] = booking_timelines.event_type_sorted.apply(book_count)

In [168]:
booking_timelines[booking_timelines.num_book > 1]

user_id     trip                from_to  distance_km duration  \
14958  690a8c  BNE_SIN  2017-07-12_2017-07-12  6147.753956   0 days   

                                                   ts_dt  \
14958  [2017-04-21 02:35:20, 2017-04-21 02:43:58, 201...   

                 event_type num_adults num_children  \
14958  [search, book, book]  [2, 2, 2]    [2, 2, 2]   

                                            ts_dt_sorted  \
14958  [2017-04-21 02:35:20, 2017-04-21 02:43:51, 201...   

          event_type_sorted num_adults_sorted num_children_sorted  \
14958  [search, book, book]         [2, 2, 2]           [2, 2, 2]   

       last_event_book  num_before  num_book  
14958                1           2         2

There is only one user who has two book events for same trip.

In [171]:
search_count_before_book = lambda x: len(x[:x.index('book')])
booking_timelines.loc[:,'num_search_before_book'] = booking_timelines.event_type_sorted.apply(search_count_before_book)


/anaconda3/lib/python3.6/site-packages/pandas/core/indexing.py:362: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[key] = _infer_fill_value(value)
/anaconda3/lib/python3.6/site-packages/pandas/core/indexing.py:543: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [174]:
booking_timelines

user_id     trip                from_to   distance_km duration  \
31     0035d4  DUS_AGP  2018-02-08_2018-02-20   1839.949396  12 days   
42     004671  DUS_MUC  2017-05-14_2017-05-15    486.185022   1 days   
49     005e14  PAR_YMQ  2017-05-07_2017-05-20   5506.315396  13 days   
57     006e35  HAM_TXL  2017-08-23_2017-08-31    246.123282   8 days   
165    011d58  BER_FRA  2017-04-28_2017-04-28    423.247353   0 days   
208    0166cf  MUC_PMI  2017-07-06_2017-07-15   1188.284804   9 days   
218    0182d0  DUS_KTM  2017-06-16_2017-07-05   6871.855604  19 days   
287    01f2a2  CGN_BER  2017-05-19_2017-05-21    470.077408   2 days   
329    022b2f  STR_ANK  2017-08-03_2017-08-17   2112.058125  14 days   
374    027919  BER_AGP  2017-10-12_2017-11-02   2240.628843  21 days   
385    028cd3  BKK_HAN  2017-05-09_2017-05-09   1003.240844   0 days   
388    028fcb  MUC_LIS  2017-09-24_2017-10-09   1963.732768  15 days   
398    02a312  NYC_CUN  2017-05-04_2017-05-12   2501.989890   8 days   
405    02b7c2  PAR_TLS  2017-05-18_2017-05-20    587.944045   2 days   
411    02c13e  DUS_ALC  2017-06-05_2017-06-17   1549.329193  12 days   
421    02cd28  BER_VIE  2017-04-22_2017-04-22    537.820574   0 days   
426    02d63d  NCE_BCN  2017-06-07_2017-06-09    485.309913   2 days   
458    031d5e  FRA_TXL  2017-05-14_2017-05-17    419.565351   3 days   
504    0388d0  HAV_MUC  2017-05-16_2017-05-16   8409.945169   0 days   
534    03c4d2  MRS_TYO  2017-11-27_2017-12-12  10084.352389  15 days   
585    040016  STR_HAM  2017-05-15_2017-05-22    536.927927   7 days   
588    0403b3  PAR_VCE  2017-08-10_2017-08-17    843.760149   7 days   
611    0426b7  MUC_VCE  2017-07-17_2017-07-17    305.603883   0 days   
650    046c18  MUC_LIS  2017-05-06_2017-05-13   1963.732768   7 days   
655    047070  BER_LED  2017-04-22_2017-05-01   1314.268755   9 days   
689    04abef  ZRH_BCN  2017-05-20_2017-05-24    843.412411   4 days   
709    04daae  DUS_AYT  2017-06-01_2017-06-10   2473.829620   9 days   
731    04ffa4  DUS_ZRH  2017-05-17_2017-05-19    438.562348   2 days   
780    057477  STR_MAD  2017-06-27_2017-07-01   1376.462502   4 days   
805    058fb2  DUS_MUC  2017-05-16_2017-05-22    486.185022   6 days   
...       ...      ...                    ...           ...      ...   
35555  fae688  DUS_BKK  2017-04-29_2017-04-29   9075.229329   0 days   
35558  fae7a7  HAM_MUC  2017-08-10_2017-08-11    614.690248   1 days   
35625  fb527d  MUC_DUS  2017-05-31_2017-05-31    486.185022   0 days   
35718  fbfb41  NCE_YVR  2017-09-24_2017-10-03   8611.903185   9 days   
35730  fc16cc  PAR_BCN  2017-05-25_2017-05-26    829.648484   1 days   
35767  fc58ae  DUS_BCN  2017-06-12_2017-06-26   1148.623804  14 days   
35771  fc5df8  DUS_LIS  2017-10-26_2017-11-09   1862.399590  14 days   
35780  fc69c9  FRA_NYC  2017-11-14_2017-11-30   6202.585984  16 days   
35804  fc941a  HAM_LON  2017-07-23_2017-07-28    726.159903   5 days   
35814  fc99b0  MUC_ARN  2017-06-21_2017-06-26   1344.107401   5 days   
35820  fca078  STR_TXL  2017-07-12_2017-07-15    510.269063   3 days   
35827  fcb6a2  BER_JNB  2017-05-19_2017-06-02   8868.076951  14 days   
35842  fcceb0  BER_OSL  2017-05-15_2017-05-21    837.856053   6 days   
35847  fcceb1  FRA_CTU  2017-05-31_2017-06-14   7799.372916  14 days   
35869  fcf1b8  STR_LON  2017-05-19_2017-05-24    731.070713   5 days   
35913  fd49f2  DUS_ZRH  2017-08-08_2017-08-11    438.562348   3 days   
35928  fd6d18  FRA_SGN  2018-01-18_2018-02-03   9645.902446  16 days   
35930  fd6fb5  PAR_PMI  2017-05-08_2017-05-09   1034.554088   1 days   
35933  fd714d  CPT_FRA  2017-11-29_2017-11-29   9400.189402   0 days   
35940  fd7cea  NCE_ORY  2017-04-27_2017-04-28    676.256264   1 days   
35970  fdb61d  FRA_ICN  2017-10-15_2017-11-23   8531.625549  39 days   
36031  fe1fbf  STR_BER  2017-06-03_2017-06-05    511.559243   2 days   
36085  feab36  MUC_BER  2017-05-16_2017-05-24    504.015245   8 days   
36101  febf52  CGN_PRG  2017-10-14_2017-10-17 